edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionReauthenticator.cs

```diff
@@ -0,0 +1,222 @@
+                                    }
+                                    else
+                                    {
+                                        Events.ServiceIdentityUpdatedValidated(serviceIdentity.Id);
+                                    }
+                                });
+                        });
+                }
+                else
+                {
+                    Events.DeviceNotConnected(serviceIdentity.Id);
+                }
+            }
+            catch (Exception ex)
+            {
+                Events.ErrorReauthenticating(serviceIdentity.Id, ex);
+            }
+        }
+
+        async void HandleServiceIdentityRemove(object sender, string id)
+        {
+            try
+            {
+                Events.ServiceIdentityRemoved(id);
+                await this.connectionManager.RemoveDeviceConnection(id);
+            }
+            catch (Exception ex)
+            {
+                Events.ErrorRemovingConnection(ex, id);
+            }
+        }
+
+        static class Events
+        {
+            static readonly ILogger Log = Logger.Factory.CreateLogger<ConnectionReauthenticator>();
+            const int IdStart = HubCoreEventIds.PeriodicConnectionAuthenticator;
+
+            enum EventIds
+            {
+                ErrorReauthenticating = IdStart,
+                ErrorRemovingConnection,
+                ServiceIdentityUpdated,
+                ServiceIdentityUpdatedRemoving,
+                ServiceIdentityUpdatedValidated,
+                ServiceIdentityRemoved,
+                ClientCredentialsResult,
+                DeviceNotConnected,
+                StartingReauthTimer
+            }
+
+            public static void ErrorReauthenticating(Exception ex)
+            {
+                Log.LogWarning((int)EventIds.ErrorReauthenticating, ex, "Error re-authenticating connected clients.");
+            }
+
+            internal static void ErrorReauthenticating(string id, Exception ex)
+            {
+                Log.LogWarning((int)EventIds.ErrorReauthenticating, ex, $"Error re-authenticating client {id}, closing the connection.");
+            }
+
+            public static void ClientCredentialsResult(IIdentity identity, bool result)
+            {
+                if (result)
+                {
+                    Log.LogDebug((int)EventIds.ClientCredentialsResult, $"Reauthenticated client {identity.Id} successfully");
+                }
+                else
+                {
+                    Log.LogWarning((int)EventIds.ClientCredentialsResult, $"Reauthenticating client {identity.Id} failed, removing client connection");
+                }
+            }
+
+            public static void ClientCredentialsNotFound(IIdentity identity)
+            {
+                Log.LogError((int)EventIds.ClientCredentialsResult, $"Unable to reauthenticate client {identity.Id} as the client credentials were not found, removing client connection");
```

Comment:
Is this really an error or a warning? It's fine if it's actually an error, just asking?

Reply:
Good catch, should be a warning.